src/librustc_mir/build/expr/into.rs

```diff
@@ -197,7 +209,7 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                         // Execute the body, branching back to the test.
                         let body_block_end = unpack!(this.into(&tmp, body_block, body));
                         this.cfg.terminate(body_block_end, source_info,
```

Comment:
So, in the new diagram above, `body_block_end` does not appear -- but I presume it is the end of the loop body? In that case, I still think that this is the wrong place for this terminator. I think we want the false unwind to be the terminator for the `loop_block`. Something like this:

```rust
if let Some(cond_expr) = opt_cond_expr {
    ... // as before
} else {
    body_block = this.cfg.start_new_block();
    let diverge_cleanup = this.diverge_cleanup();
    this.cfg.terminate(
         loop_block,
         TerminatorKind::FalseUnwind { real_target: body_block, unwind: Some(diverge_cleanup) }
    );
}
```

then there is no need for the `out_terminator` variable.

In particular I think that will be important for this test case, which would be good to add:

```rust
struct Foo { x: &'static u32 }

fn foo() {
    let a = 3;
    let foo = Foo { x: &a }; //~ ERROR E0597
    loop { continue; } // <-- note the continue here =)
}

fn main() { }
```